File: legacy/picogpt.py
v,u,t,l='lm_head','wpe','wte','input.txt'
s,R,M,L,F,B=isinstance,zip,sum,print,len,range
import os,math as N,random as S
S.seed(42)
if not os.path.exists(l):import urllib.request as P;P.urlretrieve('https://raw.githubusercontent.com/karpathy/makemore/refs/heads/master/names.txt',l)
O=[x.strip()for x in open(l).read().splitlines()if x.strip()];S.shuffle(O);L(f"num docs: {F(O)}")
T=sorted(set(''.join(O)));U,V=F(T),F(T)+1;L(f"vocab size: {V}")
class E:
	def __init__(A,data,children=(),local_grads=()):A.data=data;A.grad=0;A._children=children;A._local_grads=local_grads
	def __add__(B,other):A=other if s(other,E)else E(other);return E(B.data+A.data,(B,A),(1,1))
	def __mul__(B,other):A=other if s(other,E)else E(other);return E(B.data*A.data,(B,A),(A.data,B.data))
	def __pow__(A,other):return E(A.data**other,(A,),(other*A.data**(other-1),))
	def log(A):return E(N.log(A.data),(A,),(1/A.data,))
	def exp(A):return E(N.exp(A.data),(A,),(N.exp(A.data),))
	def relu(A):return E(max(0,A.data),(A,),(float(A.data>0),))
	def __neg__(A):return A*-1
	__radd__=__add__;__rmul__=__mul__
	def __sub__(A,other):return A+-other
	def __rsub__(A,other):return other+-A
	def __truediv__(A,other):return A*other**-1
	def __rtruediv__(A,other):return other*A**-1
	def backward(B):
		C,D=[],set()
		def E(v):
			if v in D:return
			D.add(v);[E(A)for A in v._children];C.append(v)
		E(B);B.grad=1
		for A in C[::-1]:
			for F,G in R(A._children,A._local_grads):F.grad+=G*A.grad
A,m,J,Y=16,4,1,8;H=A//m
G=lambda nout,nin,std=.02:[[E(S.gauss(0,std))for _ in B(nin)]for _ in B(nout)]
C={t:G(V,A),u:G(Y,A),v:G(V,A)}
for D in B(J):
	p=f"layer{D}."
	C[p+'attn_wq'],C[p+'attn_wk'],C[p+'attn_wv'],C[p+'attn_wo'],C[p+'mlp_fc1'],C[p+'mlp_fc2']=G(A,A),G(A,A),G(A,A),G(A,A,0),G(4*A,A),G(A,4*A,0)
W=[C for A in C.values()for B in A for C in B];L(f"num params: {F(W)}")
def I(x,w):return[M(a*b for a,b in R(r,x))for r in w]
def Z(a):m=max(x.data for x in a);a=[(x-m).exp()for x in a];s=M(a);return[x/s for x in a]
def a(x):s=(M(i*i for i in x)/F(x)+1e-5)**-.5;return[i*s for i in x]
def n(token_id,pos_id,keys,values):
	A=a([i+j for i,j in R(C[t][token_id],C[u][pos_id])])
	for D in B(J):
		G=A;A=a(A);q,k,v0=I(A,C[f"layer{D}.attn_wq"]),I(A,C[f"layer{D}.attn_wk"]),I(A,C[f"layer{D}.attn_wv"]);keys[D].append(k);values[D].append(v0);L0=[]
		for V0 in B(m):
			e=V0*H;w=q[e:e+H];k0=[x[e:e+H]for x in keys[D]];v1=[x[e:e+H]for x in values[D]];p=Z([M(w[i]*k0[j][i]for i in B(H))/H**.5 for j in B(F(k0))]);L0.extend([M(p[j]*v1[j][i]for j in B(F(v1)))for i in B(H)])
		A=[i+j for i,j in R(I(L0,C[f"layer{D}.attn_wo"]),G)];G=A;A=I(a(A),C[f"layer{D}.mlp_fc1"]);A=[x.relu()**2 for x in A];A=[i+j for i,j in R(I(A,C[f"layer{D}.mlp_fc2"]),G)]
	return I(A,C[v])
x,b,c,y=.01,.9,.95,1e-8;d,e=[0.]*F(W),[0.]*F(W);f=500
for P in B(f):
	z=O[P%F(O)];g=[U,*[T.index(A)for A in z],U];o=min(Y,F(g)-1);h,i=[[]for _ in B(J)],[[]for _ in B(J)];p=[]
	for Q in B(o):
		K,A0=g[Q:Q+2];k=Z(n(K,Q,h,i));p.append(-k[A0].log())
	q=M(p)/o;q.backward();A2=x*.5*(1+N.cos(N.pi*P/f))
	for D,X in enumerate(W):
		d[D]=b*d[D]+(1-b)*X.grad;e[D]=c*e[D]+(1-c)*X.grad**2;A3=d[D]/(1-b**(P+1));A4=e[D]/(1-c**(P+1));X.data-=A2*A3/(A4**.5+y);X.grad=0
	L(f"step {P+1:4d} / {f:4d} | loss {q.data:.4f}")
A5=.5;L('\n--- inference ---')
for A6 in B(20):
	h,i=[[]for _ in B(J)],[[]for _ in B(J)];K=U;r=[]
	for Q in B(Y):
		K=S.choices(B(V),weights=[A.data for A in Z([A/A5 for A in n(K,Q,h,i)])])[0]
		if K==U:break
		r.append(T[K])
	L(f"sample {A6+1:2d}: {''.join(r)}")

File: picogpt.js
#!/usr/bin/env node
// PicoGPT.js — GPT trained from scratch in JavaScript
// Port of picogpt.py by @kuberwastaken
// Custom autograd · multi-head attention · AdamW · training & inference

import { readFileSync, writeFileSync, existsSync } from 'fs';

// ── Seeded PRNG (xoshiro128**) ──────────────────────────────────────────────
const _s = new Uint32Array(4);
const seed = v => { let s = v|0; for (let i = 0; i < 4; i++) _s[i] = (s = Math.imul(s,1664525)+1013904223) >>> 0; };
const rand = () => { const r = Math.imul(_s[1],5), t = _s[1]<<9; _s[2]^=_s[0]; _s[3]^=_s[1]; _s[1]^=_s[2]; _s[0]^=_s[3]; _s[2]^=t; _s[3]=(_s[3]<<11)|(_s[3]>>>21); return (r>>>0)/4294967296; };
const gauss = (m,s) => m + s * Math.sqrt(-2*Math.log(rand())) * Math.cos(2*Math.PI*rand());
const shuffle = a => { for (let i = a.length-1; i > 0; i--) { const j = rand()*(i+1)|0; [a[i],a[j]]=[a[j],a[i]]; } };
const choices = (p,w) => { let t=0; for (const x of w) t+=x; let r=rand()*t; for (let i=0;i<p.length;i++) { r-=w[i]; if(r<=0) return p[i]; } return p.at(-1); };

seed(42);

// ── Data ─────────────────────────────────────────────────────────────────────
const INPUT = 'input.txt';
if (!existsSync(INPUT)) {
  console.log('Downloading names.txt …');
  const r = await fetch('https://raw.githubusercontent.com/karpathy/makemore/refs/heads/master/names.txt');
  writeFileSync(INPUT, await r.text());
}
const docs = readFileSync(INPUT,'utf-8').trim().split('\n').map(s=>s.trim()).filter(Boolean);
shuffle(docs);
console.log(`num docs: ${docs.length}`);

const chars = [...new Set(docs.join(''))].sort();
const BOS = chars.length, V = chars.length + 1;
console.log(`vocab size: ${V}`);

// ── Autograd ─────────────────────────────────────────────────────────────────
let _gen = 0;
class Value {
  constructor(d, ch=[], lg=[]) { this.data=d; this.grad=0; this._ch=ch; this._lg=lg; this._gen=0; }
  add(o) { o = o instanceof Value ? o : new Value(o); return new Value(this.data+o.data,[this,o],[1,1]); }
  mul(o) { o = o instanceof Value ? o : new Value(o); return new Value(this.data*o.data,[this,o],[o.data,this.data]); }
  pow(n) { return new Value(this.data**n,[this],[n*this.data**(n-1)]); }
  log()  { return new Value(Math.log(this.data),[this],[1/this.data]); }
  exp()  { const e=Math.exp(this.data); return new Value(e,[this],[e]); }
  relu() { return new Value(Math.max(0,this.data),[this],[+(this.data>0)]); }
  neg()  { return this.mul(-1); }
  sub(o) { return this.add(o instanceof Value ? o.neg() : -o); }
  div(o) { return this.mul(o instanceof Value ? o.pow(-1) : 1/o); }
  backward() {
    const g=++_gen, ord=[];
    (function tp(v){ if(v._gen===g) return; v._gen=g; for(const c of v._ch) tp(c); ord.push(v); })(this);
    this.grad=1;
    for (let i=ord.length-1; i>=0; i--) { const v=ord[i]; for (let j=0;j<v._ch.length;j++) v._ch[j].grad += v._lg[j]*v.grad; }
  }
}

// ── Model ────────────────────────────────────────────────────────────────────
const D=16, M=4, J=1, Y=8, H=D/M;
const mat = (r,c,std=.02) => Array.from({length:r}, ()=>Array.from({length:c}, ()=>new Value(gauss(0,std))));
const W = { wte:mat(V,D), wpe:mat(Y,D), lm_head:mat(V,D) };
for (let i=0; i<J; i++) {
  W[`l${i}.wq`]=mat(D,D); W[`l${i}.wk`]=mat(D,D); W[`l${i}.wv`]=mat(D,D);
  W[`l${i}.wo`]=mat(D,D,0); W[`l${i}.f1`]=mat(4*D,D); W[`l${i}.f2`]=mat(D,4*D,0);
}
const params = Object.values(W).flat(Infinity);
console.log(`num params: ${params.length}`);

// ── Ops ──────────────────────────────────────────────────────────────────────
const sum   = a => a.reduce((x,y) => x.add(y));
const linear = (x,w) => w.map(r => sum(r.map((v,i) => v.mul(x[i]))));
const softmax = a => { const m=Math.max(...a.map(v=>v.data)); const e=a.map(v=>v.sub(m).exp()); const s=sum(e); return e.map(v=>v.div(s)); };
const rmsnorm = a => { const s=sum(a.map(v=>v.mul(v))).mul(1/a.length).add(1e-5).pow(-.5); return a.map(v=>v.mul(s)); };

// ── Forward ──────────────────────────────────────────────────────────────────
function fwd(tok, pos, keys, vals) {
  let x = rmsnorm(W.wte[tok].map((v,i) => v.add(W.wpe[pos][i])));
  for (let l=0; l<J; l++) {
    let res=x; x=rmsnorm(x);
    const q=linear(x,W[`l${l}.wq`]), k=linear(x,W[`l${l}.wk`]), v0=linear(x,W[`l${l}.wv`]);
    keys[l].push(k); vals[l].push(v0);
    const heads=[];
    for (let h=0; h<M; h++) {
      const o=h*H, qh=q.slice(o,o+H);
      const kh=keys[l].map(k=>k.slice(o,o+H)), vh=vals[l].map(v=>v.slice(o,o+H));
      const attn=softmax(kh.map(k=>sum(qh.map((q,i)=>q.mul(k[i]))).mul(1/H**.5)));
      for (let i=0;i<H;i++) heads.push(sum(attn.map((a,j)=>a.mul(vh[j][i]))));
    }
    x=linear(heads,W[`l${l}.wo`]).map((v,i)=>v.add(res[i]));
    res=x; x=rmsnorm(x);
    x=linear(x,W[`l${l}.f1`]).map(v=>v.relu().pow(2));
    x=linear(x,W[`l${l}.f2`]).map((v,i)=>v.add(res[i]));
  }
  return linear(x, W.lm_head);
}

// ── Training ─────────────────────────────────────────────────────────────────
const lr=.01, b1=.9, b2=.95, eps=1e-8, steps=500;
const mb=new Float64Array(params.length), vb=new Float64Array(params.length);

for (let step=0; step<steps; step++) {
  const doc=docs[step%docs.length];
  const toks=[BOS,...Array.from(doc,c=>chars.indexOf(c)),BOS];
  const ctx=Math.min(Y,toks.length-1);
  const keys=Array.from({length:J},()=>[]), vals=Array.from({length:J},()=>[]);
  const losses=[];
  for (let t=0;t<ctx;t++) { const logits=softmax(fwd(toks[t],t,keys,vals)); losses.push(logits[toks[t+1]].log().neg()); }
  const loss=sum(losses).mul(1/ctx);
  loss.backward();
  const lr_t=lr*.5*(1+Math.cos(Math.PI*step/steps));
  for (let i=0;i<params.length;i++) {
    const p=params[i];
    mb[i]=b1*mb[i]+(1-b1)*p.grad; vb[i]=b2*vb[i]+(1-b2)*p.grad**2;
    p.data -= lr_t*(mb[i]/(1-b1**(step+1))) / ((vb[i]/(1-b2**(step+1)))**.5+eps);
    p.grad=0;
  }
  console.log(`step ${String(step+1).padStart(4)} / ${String(steps).padStart(4)} | loss ${loss.data.toFixed(4)}`);
}

// ── Inference ────────────────────────────────────────────────────────────────
const temp=.5, ids=Array.from({length:V},(_,i)=>i);
console.log('\n--- inference ---');
for (let s=0; s<20; s++) {
  const keys=Array.from({length:J},()=>[]), vals=Array.from({length:J},()=>[]);
  let tok=BOS; const name=[];
  for (let t=0;t<Y;t++) {
    tok=choices(ids, softmax(fwd(tok,t,keys,vals).map(v=>v.div(temp))).map(v=>v.data));
    if (tok===BOS) break;
    name.push(chars[tok]);
  }
  console.log(`sample ${String(s+1).padStart(2)}: ${name.join('')}`);
}
